lib/gossip/syncer.js

```diff
@@ -0,0 +1,192 @@
+};
+
+Syncer.prototype.sync = function sync(callback) {
+    var self = this;
+    callback = callback || noop;
+
+    if (this.isSyncing) {
+        this.ringpop.logger.debug('ringpop syncer already syncing', {
+            local: this.ringpop.whoami()
+        });
+        cleanup(new events.SyncerAlreadySyncingEvent());
+        return;
+    }
+
+    this.isSyncing = true;
+    this.emit('event', new events.SyncerSyncingEvent());
+
+    // Lazily initialize by waiting until sync is started
+    if (!this.memberIterator) {
+        this.memberIterator = new MemberIterator(this.ringpop);
+    }
+
+    if (!this.ringpop.dissemination.isEmpty()) {
+        this.ringpop.logger.debug('ringpop syncer syncing skipped; changes pending', {
+            local: this.ringpop.whoami()
+        });
+        cleanup(new events.ChangesPendingEvent());
+        return;
+    }
+
+    var pingableMember = this.memberIterator.next();
+    if (!pingableMember) {
+        this.ringpop.logger.debug('ringpop syncer syncing skipped; next member does not exist', {
+            local: this.ringpop.whoami()
+        });
+        cleanup(new events.NoSyncTargetEvent());
+        return;
+    }
+
+    this.ringpop.stat('increment', 'sync.send.attempt');
+    this.ringpop.logger.info('ringpop syncer starting', {
+        local: this.ringpop.whoami(),
+        target: pingableMember.address
+    });
+    var membershipChecksum = this.ringpop.membership.checksum;
+    var startTime = Date.now();
+    this.ringpop.client.protocolSync(
+        pingableMember.address,
+        new SyncRequestHeaders(this.ringpop.config.get('syncGzipEnabled')),
+        new SyncRequestBody(membershipChecksum),
+        onSync);
+
+    function onSync(err, response) {
+        self.ringpop.stat('timing', 'sync.response', Date.now() - startTime);
+
+        if (err) {
+            self.ringpop.stat('increment', 'sync.send.error');
+            self.ringpop.logger.warn('ringpop protocol sync error', {
+                local: self.ringpop.whoami(),
+                target: pingableMember.address,
+                err: err
+            });
+            cleanup(new events.SyncFailedEvent());
+            return;
+        }
+
+        self.ringpop.stat('increment', 'sync.send.success');
+        var membershipChanges = response.membershipChanges;
+        if (!Array.isArray(membershipChanges) || membershipChanges.length === 0) {
+            self.ringpop.stat('increment', 'sync.changes.none');
+            self.ringpop.logger.info('ringpop syncer synced; no changes', {
+                local: self.ringpop.whoami(),
+                target: pingableMember.address,
+                currentChecksum: self.ringpop.membership.checksum,
+                targetChecksum: response.membershipChecksum
+            });
+            cleanup(new events.SyncEmptyEvent());
+            return;
+        }
+
+        var updates = self.ringpop.membership.update(membershipChanges);
```

Comment:
Agreed, but I'm going to punt on this for this PR. It's an optimization that we can add later.